FirebaseAppDistribution/Sources/FIRAppDistributionAuthPersistence.m

```diff
@@ -16,77 +16,99 @@
 NS_ASSUME_NONNULL_BEGIN
 
+NSString *const kFIRAppDistributionKeychainErrorDomain = @"com.firebase.app_distribution.internal";
+
 @implementation FIRAppDistributionAuthPersistence
 
-+ (BOOL)clearAuthState {
-    NSMutableDictionary *keychainQuery = [self getKeyChainQuery];
-    OSStatus status = SecItemDelete((CFDictionaryRef)keychainQuery);
-    
-    if (status != errSecSuccess && status != errSecItemNotFound) {
-        NSLog(@"AUTH ERROR. Cant delete auth state in keychain");
-    } else {
-        NSLog(@"AUTH SUCCESS! deleted auth state in the keychain");
-    }
-    return errSecSuccess || status == errSecItemNotFound;
++ (BOOL)clearAuthState:(NSError **_Nullable)error {
+  NSMutableDictionary *keychainQuery = [self getKeyChainQuery];
+  OSStatus status = SecItemDelete((CFDictionaryRef)keychainQuery);
+
+  if (status != errSecSuccess && status != errSecItemNotFound && error) {
+    NSString *desc = NSLocalizedString(
+        @"Failed to clear auth state from keychain. Tester will overwrite data on sign in.",
+        @"Error message for failure to retrieve auth state from keychain");
+    NSDictionary *userInfo = @{NSLocalizedDescriptionKey : desc};
+    *error = [NSError errorWithDomain:kFIRAppDistributionKeychainErrorDomain
+                                 code:FIRAppDistributionErrorTokenDeletionFailure
+                             userInfo:userInfo];
+  }
+  return errSecSuccess || status == errSecItemNotFound;
 }
 
-+ (OIDAuthState*)retrieveAuthState {
-    NSMutableDictionary *keychainQuery = [self getKeyChainQuery];
-      [keychainQuery setObject:(id)kCFBooleanTrue forKey:(id)kSecReturnData];
-      [keychainQuery setObject:(id)kSecMatchLimitOne forKey:(id)kSecMatchLimit];
-      CFDataRef passwordData = NULL;
-      NSData *result = nil;
-      OSStatus status = SecItemCopyMatching((CFDictionaryRef)keychainQuery,
-                                         (CFTypeRef *)&passwordData);
-      if (status == noErr && 0 < [(__bridge NSData *)passwordData length]) {
-        result = [(__bridge NSData *)passwordData copy];
-      } else {
-          NSLog(@"AUTH ERROR - cannot lookup keystore config");
-      }
-      if (passwordData != NULL) {
-        CFRelease(passwordData);
-      }
-    
-    OIDAuthState *authState = nil;
-    if(result) {
-        authState = (OIDAuthState *)[NSKeyedUnarchiver unarchiveObjectWithData:result];
-    }
-    
-    return authState;
++ (OIDAuthState *)retrieveAuthState:(NSError **_Nullable)error {
+  NSMutableDictionary *keychainQuery = [self getKeyChainQuery];
+  [keychainQuery setObject:(id)kCFBooleanTrue forKey:(id)kSecReturnData];
+  [keychainQuery setObject:(id)kSecMatchLimitOne forKey:(id)kSecMatchLimit];
+  CFDataRef passwordData = NULL;
+  NSData *result = nil;
+  OSStatus status = SecItemCopyMatching((CFDictionaryRef)keychainQuery, (CFTypeRef *)&passwordData);
+
+  if (status == noErr && 0 < [(__bridge NSData *)passwordData length]) {
+    result = [(__bridge NSData *)passwordData copy];
```

Comment:
So it's weird, but make passwordData a `NSData *`, and cast it to `void *` when passing to `SecItemCopyMatching`. https://stackoverflow.com/questions/17822895/how-to-cast-an-indirect-pointer-in-objective-c

Apple's toll-free bridge is only one way here it seems

Reply:
In general it's nicer to stick with the NS- APIs than the CF ones so you don't need to deal with memory management

Reply:
👍